### ISSUE TITLE
stmgr hung due to zookeeper client connection lost

### ISSUE THREAD

**Issue by huijunw** [apache/incubator-heron#2235]
We found a stmgr hung for long time yesterday in a topology.
The stmgr log showed before it got rescheduled:
```
I0822 22:09:22.767264 60484 zkclient.cpp:239] Getting zknode /storm/heron/states/metricscaches/summingbird_tsar_VideoAnalytics
I0822 22:09:22.780617 60495 zkclient.cpp:59] ZKClient CallWatcher called with type SESSION_EVENT and state EXPIRED_SESSION_STATE
I0822 22:09:22.780640 60495 zkclient.cpp:59] ZKClient CallWatcher called with type SESSION_EVENT and state EXPIRED_SESSION_STATE
I0822 22:09:22.780644 60495 zkclient.cpp:307] ZKClient GlobalWatcher called with type SESSION_EVENT and state EXPIRED_SESSION_STATE
I0822 22:09:22.780699 60484 heron-zkstatemgr.cpp:323] Received an event, Type: SESSION_EVENT, State: EXPIRED_SESSION_STATE
I0822 22:09:22.780709 60484 heron-zkstatemgr.cpp:332] Deleting current zk client...
(file end)
```
This is the last several lines of the stmgr log before the container died.

https://github.com/twitter/heron/blob/master/heron/statemgrs/src/cpp/statemgr/heron-zkstatemgr.cpp#L352 is the last line executed and then the stmgr hung and there were no more logs. However heron-executor thought the stmgr was still running and did not restart it.

Then the heron-instance still sent message to stmgr and failed, which is shown in the heron-instance log, and heron-instance died and heron-executor restarted the heron-instance too many times and heron-executor quited. Finally Aurora rescheduled the container and the topology recovered. The whole process costed more than 1 hour.

@kramasamy I found your notes in the code saying it could block the execution https://github.com/twitter/heron/blob/master/heron/statemgrs/src/cpp/statemgr/heron-zkstatemgr.cpp#L353. Could you explain more on this?

**Comment by srkukarni** [comment]
One potential reason is that our zk library is too old and contains many bugs which have since been solved. Here is the change to update that to latest stable version
https://github.com/twitter/heron/pull/2245

**Comment by huijunwu** [comment]
should be fixed in #2908, cannot reproduce in the recent releases
